python/demo/demo_mixed-poisson.py

```diff
@@ -109,88 +104,296 @@
+# types
+dtype = dolfinx.default_scalar_type
+xdtype = dolfinx.default_real_type
+# -
 
-msh = mesh.create_unit_square(MPI.COMM_WORLD, 32, 32, mesh.CellType.quadrilateral)
+# Create mesh a mesh in two dimensions. The iterative solver constructed
+# later requires special construction that is specific to two
+# dimensions. Application in three-dimensions would require a number of
+# changes to the linear solver.
 
+# +
+msh = create_unit_square(MPI.COMM_WORLD, 48, 48, CellType.quadrilateral, dtype=xdtype)
+# -
+#
+# Here we construct compatible function spaces for the mixed Poisson
+# problem. The `V` Raviart-Thomas ($\mathbb{RT}$) space is a vector-valued $H({\rm
+# div})$ conforming space. The `W` space is a space of discontinuous
+# Lagrange function of degree `k`.
+# ```{note}
+# The $\mathbb{RT}_{k}$ element in DOLFINx/Basix is usually denoted as
+# $\mathbb{RT}_{k-1}$ in the literature.
+# ```
+# In the lowest-order case $k=1$. It can be increased, by the
+# convergence of the iterative solver will degrade.
+
+# +
 k = 1
-Q_el = element("BDMCF", msh.basix_cell(), k, dtype=default_real_type)
-P_el = element("DG", msh.basix_cell(), k - 1, dtype=default_real_type)
-V_el = mixed_element([Q_el, P_el])
-V = fem.functionspace(msh, V_el)
+V = fem.functionspace(msh, element("RT", msh.basix_cell(), k, dtype=xdtype))
+W = fem.functionspace(msh, element("Discontinuous Lagrange", msh.basix_cell(), k - 1, dtype=xdtype))
+# -
+
+# Trial functions for $\sigma$ and $u$ are declared on the space $V$ and
+# $W$, with corresponding test functions $\tau$ and $v$:
+
+# +
+(sigma, u) = TrialFunction(V), TrialFunction(W)
+(tau, v) = TestFunction(V), TestFunction(W)
+# -
 
-(sigma, u) = TrialFunctions(V)
-(tau, v) = TestFunctions(V)
+# The source function is set to be $f = 10\exp(-((x_{0} - 0.5)^2 +
+# (x_{1} - 0.5)^2) / 0.02)$:
 
+# +
 x = SpatialCoordinate(msh)
-f = 10.0 * exp(-((x[0] - 0.5) * (x[0] - 0.5) + (x[1] - 0.5) * (x[1] - 0.5)) / 0.02)
+f = 10 * exp(-((x[0] - 0.5) * (x[0] - 0.5) + (x[1] - 0.5) * (x[1] - 0.5)) / 0.02)
+# -
+
+# We now declare the blocked bilinear and linear forms. The rows of `a`
+# and `L` correspond to the $\tau$ and $v$ test functions, respectively.
+# The columns of `a` correspond to the $\sigma$ and $u$ trial functions,
+# respectively. Note that `a[1][1]` is empty, which is denoted by
+# `None`. This zero block is typical of a saddle-point problem. In the
+# `L[0]` block, the test function $\tau$ is multiplied by a zero
+# `Constant`, which is evaluated at runtime. We do this to preserve
+# knowledge of the test space in the block. *Note that the defined `L`
+# corresponds to $u_{0} = 0$ on $\Gamma_{D}$.*
 
+# +
 dx = Measure("dx", msh)
-a = inner(sigma, tau) * dx + inner(u, div(tau)) * dx + inner(div(sigma), v) * dx
-L = -inner(f, v) * dx
+a = [[inner(sigma, tau) * dx, inner(u, div(tau)) * dx], [inner(div(sigma), v) * dx, None]]
+L = [
+    inner(fem.Constant(msh, np.zeros(2, dtype=dtype)), tau) * dx,
+    -inner(f, v) * dx,
+]
```

Comment:
Is there a reason for not using `MixedFunctionSpace` here, so that the user doesn't have to manually specify the block structure, and can use `ufl.extract_blocks` to get the blocked matrix and vector, as done in https://github.com/FEniCS/dolfinx/pull/3508, ref: https://docs.fenicsproject.org/dolfinx/v0.9.0/python/demos/demo_hdg.html

Reply:
I found `MixedFunctionSpace` confusing - sometimes we use it, sometimes we don't. I was going to register a separate issue on it. Why do we have two ways of doing the same thing? Confusion compounded by not always using the `ufl` namespace in demos to clearly distinguish between UFL and DOLFINx functions/classes (see #3532).

Reply:
It has not been consistently used as it was introduced/fixed right before the release of 0.9.0.
One can do things in two ways because:
1. One want the flexibility to write out the blocks yourself, and the input to assemble_matrix_block/nest is just a list/nested list of variational forms.
2. Using `MixedFunctionSpace` and `ufl.extract_blocks` massively simplifies the readability of the variational form, especially for problems with more than two blocks. (I think @jpdean has some good examples of this in: https://github.com/jpdean/mixed_domain_demos). I also use `extract_blocks` alot in my own projects, as I have more than just two blocks, and filling out all the `None` blocks consistently is utterly painful and error-prone.

Reply:
This also implicitly links to: https://github.com/FEniCS/ufl/pull/303#issuecomment-2575598818
which would give us yet another option as to how to handle mixed/sub-mesh stuff.